File: Tests_tmp/ImageCompressionLib/Containers/ModuleOPCTest.java
package ImageCompressionLib.Containers;

import ImageCompressionLib.Constants.State;
import ImageCompressionLib.ProcessingModules.ModuleOPC.ModuleOPC;
import org.junit.Before;
import org.junit.Test;

import java.util.Date;

import static org.junit.Assert.*;

public class ModuleOPCTest {
    int size=200;
    Flag flag=new Flag("0");
    TripleShortMatrix tripleShortMatrix =new TripleShortMatrix(size,size, State.RGB);
    @Before
    public void setUp() throws Exception {
        for(int i=0;i<size;i++){
            for(int j=0;j<size;j++){
                tripleShortMatrix.getA()[i][j]=(short)(i+j);
                tripleShortMatrix.getB()[i][j]=(short)(i+j);
                tripleShortMatrix.getC()[i][j]=(short)(i+j);
            }
        }
    }

    @Test
    public void TestDefault(){
        TripleDataOpcMatrix tripleDataOpcMatrix =new ModuleOPC(tripleShortMatrix,flag,true).getBoxOfOpc(true);
        DataOpc[][] a,b,c;
        ModuleOPC moduleOPC1 =new ModuleOPC(tripleDataOpcMatrix, flag,true);
        TripleShortMatrix res= moduleOPC1.getMatrix(true);

        AssertMatrixInRange(tripleShortMatrix,res,1,true);
    }
    @Test
    public void TestTime(){
        ModuleOPC moduleOPC =new ModuleOPC(tripleShortMatrix,flag,true);
        Date t1=new Date();
        moduleOPC.directOPCMultiThreads();
        moduleOPC.reverseOPCMultiThreads();
        Date t2=new Date();
        System.out.println("in multi threads t="+(t2.getTime()-t1.getTime()));

        Date t3=new Date();
        moduleOPC.directOPC();
        moduleOPC.reverseOPC();
        Date t4=new Date();
        System.out.println("in one thread t="+(t4.getTime()-t3.getTime()));

        Date t5=new Date();
        moduleOPC.directOpcGlobalBase(5,5);
        moduleOPC.reverseOPC();
        Date t6=new Date();
        System.out.println("Global Base t="+(t6.getTime()-t5.getTime()));

        assertTrue((t4.getTime()-t3.getTime())>(t2.getTime()-t1.getTime()));
    }

    @Test
    public void TestAssertFun(){
        TripleShortMatrix a=new TripleShortMatrix(size,size,State.RGB);
        TripleShortMatrix b=new TripleShortMatrix(size,size,State.RGB);
        AssertMatrixInRange(a,b,0,true);

        a.getA()[0][0]=1;
        AssertMatrixInRange(a,b,2,true);

        a.getA()[0][0]=200;
        a.getC()[0][0]=200;
        a.getB()[0][0]=200;
        AssertMatrixInRange(a,b,1,false);
    }

    private void AssertMatrixInRange(TripleShortMatrix m1, TripleShortMatrix m2, int delta, Boolean inRange){
        AssertArrayArrayInRange(m1.getA(), m2.getA(),delta,inRange);
        AssertArrayArrayInRange(m1.getB(), m2.getB(),delta,inRange);
        AssertArrayArrayInRange(m1.getC(), m2.getC(),delta,inRange);
    }
    private void AssertArrayArrayInRange(short[][] a,short[][] b,int delta,boolean inRange){
        boolean total=true;
        for(int i=0;i<a.length;i++){
            for(int j=0;j<a[0].length;j++){
                boolean isEqual=false;
                for(int d=-delta;d<=delta;d++){
                    if(a[i][j]==b[i][j]+d)
                        isEqual=true;
                }
                if(inRange)
                    assertTrue(isEqual);
                total=total&&isEqual;
            }
        }
        if(inRange)
            assertTrue(total);
        else
            assertFalse(total);
    }


}

File: Tests_tmp/ImageCompressionLib/processing_modules/ModuleOpcTest.kt
package ImageCompressionLib.processing_modules

import ImageCompressionLib.constants.State
import ImageCompressionLib.containers.EncryptParameters
import ImageCompressionLib.containers.type.ByteVector
import ImageCompressionLib.containers.type.Size
import ImageCompressionLib.utils.functions.ImageIOTest
import ImageCompressionLib.utils.functions.opc.IStegoMessageUtil
import ImageCompressionLib.utils.objects.TimeManager
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import java.util.*

@RunWith(Parameterized::class)
class ModuleOpcTest(val w: Int,val h: Int,val loop: Int) {
    /*@Test
    fun TestDirectTime(){
        val m=ImageIOTest.createMatrix(1080,1920)
        m.state=State.DCT
        val opcModule= ModuleOpc(m, Parameters.createParametresForTest())

        val loop=1
        val t1=Date().time
        for (i in 0..loop){
            opcModule.directOPC()
        }
        val t2=Date().time

        val t3=Date().time
        for (i in 0..loop){
            opcModule.directOPCMultiThreads()
        }
        val t4=Date().time

        System.out.println("directOPC ${loop}: one=${t2-t1}, multi=${t4-t3}")
        assertTrue((t2-t1)>(t4-t3))
    }
    @Test
    fun TestReverseTime(){
        val m=ImageIOTest.createMatrix(1080,1920)
        m.state=State.DCT
        val opcModule= ModuleOpc(m, Flag(), true)
        val box=opcModule.getBoxOfOpc(true)

        val loop=1
        val t1=Date().time
        for (i in 0..loop){
            opcModule.reverseOPC()
        }
        val t2=Date().time
//        println("Start async")
        val t3=Date().time
        for (i in 0..loop){
            opcModule.reverseOPCMultiThreads()
        }
        val t4=Date().time

        System.out.println("reverseOPC ${loop}: one=${t2-t1}, multi=${t4-t3}")
        assertTrue((t2-t1)>(t4-t3))
    }*/
    @Test
    fun GlobalTest(){
        val m=ImageIOTest.createMatrix(w,h)
        m.state=State.Dct
        val opcModule= ModuleOpc(m)
        val tm=TimeManager.Instance

        val cpy=m.copy()
        assertEquals(m,cpy)

        tm.startNewTrack("ModuleOPC")
        for(i in 0 until loop) {
            tm.append("start")
            val tdom = opcModule.getTripleDataOpcMatrix(null)
            val m2 = ModuleOpc(tdom)
            val res = m2.getTripleShortMatrix(null).first
            tm.append("end")
            assertEquals(res, cpy)
        }
        println(tm.getInfoInSec())
    }
    @Test
    fun GlobalTestUnitSize7x6() {
        val m = ImageIOTest.createMatrix(w, h,Size(7, 6))
//        m.parameters.unitSize = Size(7, 6)
        m.state = State.Dct
        val cpy = m.copy()

        val opcModule = ModuleOpc(m)
        val tdom = opcModule.getTripleDataOpcMatrix(null)
        val m2 = ModuleOpc(tdom)
        val res = m2.getTripleShortMatrix(null).first

        assertEquals(res, cpy)
    }
    @Test
    fun StegoAtZero(){
        TestSteganographyAtPos(0)
    }
    @Test
    fun StegoAtFirst(){
        TestSteganographyAtPos(1)
    }
    fun TestSteganographyAtPos(pos:Int) {
        val m = ImageIOTest.createMatrix(w, h,Size(1,8))
        m.state = State.Dct
        val cpy = m.copy()
        val eP=EncryptParameters()
        eP.message= ByteVector(w/2)
        for(i in 0 until eP.message!!.maxSize)
            eP.message!!.append(Random().nextLong().toShort())
        val messCpy=eP.message!!.copy()
        eP.steganography=EncryptParameters.StegaParameters(pos){object :IStegoMessageUtil{
            override fun isUseNextBlock(): Boolean {
                return true
            }
        }}


        val opcModule = ModuleOpc(m)
        val tdom = opcModule.getTripleDataOpcMatrix(eP)
        val m2 = ModuleOpc(tdom)
        val res = m2.getTripleShortMatrix(eP)
        val mess=res.second
        val mat=res.first

        assertEquals(messCpy,mess!!)
//        cpy.assertMatrixInRange(mat,2)
    }
    companion object {
        @JvmStatic
        @Parameterized.Parameters(name = "w:{0},h:{1},loop:{2}")
        fun data(): Collection<Array<Any>> {
            return listOf(
                    arrayOf(36,24,1) as Array<Any>
                    ,arrayOf(13,25,1) as Array<Any>
                    ,arrayOf(360,240,1) as Array<Any>//477
                    ,arrayOf(360,240,3) as Array<Any>//1386
                    ,arrayOf(1920,1080,1) as Array<Any>//11343
                    ,arrayOf(1920,1080,3) as Array<Any>//34478
            )
        }
    }
}